drivers/video/video_stm32_dcmipp.c

```diff
@@ -0,0 +1,1255 @@
+}
+
+static int stm32_dcmipp_dequeue(const struct device *dev, enum video_endpoint_id ep,
+				struct video_buffer **vbuf, k_timeout_t timeout)
+{
+	struct stm32_dcmipp_pipe_data *pipe = dev->data;
+
+	if (!vbuf) {
+		return -EINVAL;
+	}
+
+	*vbuf = k_fifo_get(&pipe->fifo_out, timeout);
+	if (!*vbuf) {
+		return -EAGAIN;
+	}
+
+	return 0;
+}
+
+/*
+ * TODO: caps aren't yet handled hence give back straight the caps given by the
+ * sensor.  Normally this should be the intersection of what the sensor produces
+ * vs what the DCMIPP can input (for pipe0) and, for pipe 1 and 2, for a given
+ * input format, generate caps based on capabilities, color conversion, decimation
+ * etc
+ */
+static int stm32_dcmipp_get_caps(const struct device *dev, enum video_endpoint_id ep,
+				 struct video_caps *caps)
+{
+	const struct stm32_dcmipp_config *config = dev->config;
+	int ret;
+
+	ret = video_get_caps(config->sensor_dev, ep, caps);
+
+	caps->min_vbuf_count = 1;
+	caps->min_line_count = LINE_COUNT_HEIGHT;
+	caps->max_line_count = LINE_COUNT_HEIGHT;
+
+	return ret;
+}
+
+static const uint8_t stm32_dcmipp_isp_dec_val[] = {
+	1, 2, 4, 8
+};
+
+static const char *const stm32_dcmipp_isp_dec_menu[] = {
+	"Disabled",
+	"1/2",
+	"1/4",
+	"1/8",
+	NULL
+};
+
+static int stm32_dcmipp_init_controls(const struct device *dev)
+{
+	struct stm32_dcmipp_data *dcmipp = dev->data;
+	struct stm32_dcmipp_ctrls *ctrls = &dcmipp->ctrls;
+	int ret;
+
+	ret = video_init_menu_ctrl(&ctrls->isp_hdec, dev, VIDEO_CID_STM32_DCMIPP_ISP_HDEC, 0,
+				   stm32_dcmipp_isp_dec_menu);
+	if (ret) {
+		return ret;
+	}
+
+	return video_init_menu_ctrl(&ctrls->isp_vdec, dev, VIDEO_CID_STM32_DCMIPP_ISP_VDEC, 0,
+				    stm32_dcmipp_isp_dec_menu);
+}
+
+static int stm32_dcmipp_set_ctrl(const struct device *dev, uint32_t id)
+{
+	struct stm32_dcmipp_pipe_data *pipe = dev->data;
+	struct stm32_dcmipp_data *dcmipp = pipe->dcmipp;
+
+	switch (id) {
+	case VIDEO_CID_STM32_DCMIPP_ISP_HDEC:
+		dcmipp->isp_dec_hratio = stm32_dcmipp_isp_dec_val[dcmipp->ctrls.isp_hdec.val];
+		break;
+	case VIDEO_CID_STM32_DCMIPP_ISP_VDEC:
+		dcmipp->isp_dec_vratio = stm32_dcmipp_isp_dec_val[dcmipp->ctrls.isp_vdec.val];
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static DEVICE_API(video, stm32_dcmipp_driver_api) = {
+	.set_format = stm32_dcmipp_set_fmt,
+	.get_format = stm32_dcmipp_get_fmt,
+	.set_stream = stm32_dcmipp_set_stream,
+	.enqueue = stm32_dcmipp_enqueue,
+	.dequeue = stm32_dcmipp_dequeue,
+	.get_caps = stm32_dcmipp_get_caps,
+	.set_ctrl = stm32_dcmipp_set_ctrl,
+};
+
+static int stm32_dcmipp_enable_clock(const struct device *dev)
+{
+	const struct stm32_dcmipp_config *config = dev->config;
+	const struct device *cc_node = DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE);
+	int err;
+
+	if (!device_is_ready(cc_node)) {
+		LOG_ERR("clock control device not ready");
+		return -ENODEV;
+	}
+
+	/* Turn on DCMIPP peripheral clock */
+	/* TODO - retrieve clock index from the DT instead of hardcode here */
```

Comment:
See https://github.com/zephyrproject-rtos/zephyr/pull/85564 for equivalent functionality